src/components/features/language-selector/LanguageSelector.tsx

```diff
@@ -11,15 +15,55 @@ const displayName = locale =>
     type: 'language',
   });
 
+const isChangeEvent = (event: SyntheticEvent): event is React.ChangeEvent<HTMLSelectElement> => {
+  return event.type === 'change';
+};
+
 export const LanguageSelector = () => {
+  const { i18n } = useTranslation();
+  const currentLocale = i18n.language;
+  const router = useRouter();
+  const currentPathname = usePathname();
+
+  const handleLocaleChange: React.EventHandler<React.SyntheticEvent> = e => {
+    let newLocale: string | undefined = undefined;
+
+    if (isChangeEvent(e)) {
+      newLocale = e.target.value;
+    }
+
+    // set cookie for next-i18n-router
+    const days = 30;
+    const date = new Date();
+    date.setTime(date.getTime() + days * 24 * 60 * 60 * 1000);
+    document.cookie = `NEXT_LOCALE=${newLocale};expires=${date.toUTCString()};path=/`;
```

Comment:
Feels strange to do this manually...

Reply:
Yeah, so as an explanation: The bug you found with the language switcher was that Next.js middleware doesn't run for cached pages, and the middleware is what automatically sets the language cookie. To get around the bug we can force set the cookie and then call `router.refresh()` to apply it.